docs/concepts/controller (#21495)
kubernetes/website: content/it/docs/concepts/architecture/controller.md

---
title: Controller
content_template: templates/concept
weight: 40
---

{{% capture overview %}}

Nella robotica e nell'automazione, un _circuito di controllo_ (_control loop_) è un un'iterazione senza soluzione di continuità che regola lo stato di un sistema.

Ecco un esempio di un circuito di controllo: il termostato di una stanza.

Quando viene impostata la temperatura, si definisce attraverso il termostato lo *stato desiderato*. L'attuale temperatura nella stanza è invece lo *stato corrente*. Il termostato agisce per portare lo stato corrente il più vicino possibile allo stato desiderato accendendo e spegnendo le apparecchiature.

{{< glossary_definition term_id="controller" length="short" >}}

{{% /capture %}}

{{% capture body %}}

## Il modello del controller

Un _controller_ monitora almeno una tipo di risorsa registrata in Kubernetes. 

> Questi [oggetti](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects) hanno una proprietà chiamata *spec* (specifica) che rappresenta lo stato desiderato. Il o i *controller(s)* per quella risorsa sono responsabili di mantenere lo stato corrente il più simile possibile rispetto allo stato desiderato.

Reviewer: In Italian the plural form of English words is still singular, so: s/controller(s)/controller/

Il _controller_ potrebbe eseguire l'azione relativa alla risorsa in questione da sé; più comunemente, in Kubernetes, un _controller_ invia messaggi all'{{< glossary_tooltip text="API server" term_id="kube-apiserver" >}} che a sua volta li rende disponibili ad altri componenti nel cluster. Di seguito troverete esempi per questo scenario.

{{< comment >}}
Alcuni controller nativi, come ad esempio l'_endpoints_ controller, agiscono su oggetti che non hanno una specifica. Per semplicità, questa pagina non entra in quel dettaglio.
{{< /comment >}}

### Controllo attraverso l'API server

Il {{< glossary_tooltip term_id="job" >}} _controller_ è un esempio di un _controller_ nativo in Kubernetes. I _controller_ nativi gestiscono lo stato interagendo con l'API server presente nel cluster.

Il Job è una risorsa di Kubernetes che lancia uno o più {{< glossary_tooltip term_id="pod" text="Pods" >}} per eseguire un lavoro (task) e poi fermarsi.

(Una volta che è stato [schedulato](/docs/concepts/scheduling-eviction/), un oggetto _Pod_ diventa parte dello stato desisderato di un dato _kubelet_).

> Quando il Job _controller_ vede un nuovo lavoro da svolgere si assicura che, da qualche parte nel cluster, i _kubelets_ anche sparsi su più nodi eseguano il numero corretto di _Pods_ necessari per eseguire il lavoro richiesto. Il Job _controller_ non esegue direttamente alcun _Pod_ o _container_ bensì chiede all'API server di creare o rimuovere i _Pods_. Altri componenti appartenenti al {{< glossary_tooltip text="control plane" term_id="control-plane" >}} reagiscono in base alle nuove informazioni (ci sono nuovi _Pods_ da creare e gestire) e cooperano al completamento del job.

Reviewer: s/kubelets/kubelet/
s/Pods/Pod/g

> Dopo che un nuovo Job è stato creato, lo stato desiderato per quel Job è il suo completamento. Il Job _controller_ fa sì che lo stato corrente per quel Job sia il più vicino possibile allo stato desiderato: creare _Pods_ che eseguano il lavoro che deve essere effettuato attraverso il Job, così che il Job sia prossimo al completamento.

Reviewer: s/Pods/Pod/

I _controller_ che interagiscono con un sistema esterno trovano il loro stato desiderato attraverso l'API server, quindi comunicano direttamente con un sistema esterno per portare il loro stato corrente più in linea possibile con lo stato desiderato

(In realtà c'è un _controller_ che scala orizzontalmente i nodi nel cluster. Vedi [Cluster autoscaling](/docs/tasks/administer-cluster/cluster-management/#cluster-autoscaling)).

## Stato desiderato versus corrente {#desiderato-vs-corrente}

Kubernetes ha una visione *cloud-native* dei sistemi, ed è in grado di gestire continue modifiche.

Il cluster viene modificato continuamente durante la sua attività ed il _circuito di controllo_ è in grado di risolvere automaticamente i possibili guasti.

Fino a che i _controller_ del cluster sono in funzione ed in grado di apportare le dovute modifiche, non è rilevante che lo stato complessivo del cluster sia o meno stabile.

## Progettazione

> Come cardine della sua progettazione, Kubernetes usa vari _controller_ ognuno dei quali è responsabile per un particolare aspetto dello stato del cluster. Più comunemente, un dato _circuito di controllo_ (_controller_) usa un tipo di risorsa per il suo stato desiderato, ed utilizza anche risorse di altro tipo per raggiungere questo stato desiderato. Per esempio il Job _controller_ tiene traccia degli oggetti di tipo _Job_ (per scoprire nuove attività da eseguire) e degli oggetti di tipo _Pod_ (questi ultimi usati per eseguire i _Jobs_, e quindi per controllare quando il loro lavoro è terminato). In questo caso, qualcos'altro crea i _Jobs_, mentre il _Job_ _controller_ crea i _Pods_.

Reviewer: s/Jobs/Job/g
s/Pods/Pod/

È utile avere semplici _controller_ piuttosto che un unico, monolitico, _circuito di controllo_. I _controller_ possono guastarsi, quindi Kubernetes è stato disegnato per gestire questa eventualità.

{{< note >}}
Ci possono essere diversi _controller_ che creato o aggiornano lo stesso tipo di oggetti. Dietro le quinte, i _controller_ di Kubernetes si preoccupano esclusivamente delle risorse (di altro tipo) collegate alla risorsa primaria da essi controllata.

> Per esempio, si possono avere _Deployments_ e _Jobs_; entrambe creano _Pods_. Il Job _controller_ non distrugge i _Pods_ creati da un _Deployment_, perché ci sono informazioni (*{{< glossary_tooltip term_id="label" text="labels" >}}*) che vengono usate dal _controller_ per distinguere i _Pods_.

Reviewer: s/Deployemnts/Deployment/
s/Jobs/Job/
s/Pods/Pod/g

{{< /note >}}

## I modi per eseguire i _controller_ {#eseguire-controller}

Kubernetes annovera un insieme di _controller_ nativi che sono in esecuzione all'interno del {{< glossary_tooltip term_id="kube-controller-manager" >}}. Questi _controller_ nativi forniscono importanti funzionalità di base.

Il Deployment _controller_ ed il Job _controller_ sono esempi di _controller_ che vengono forniti direttamente da Kubernetes stesso (ovvero _controller_ "nativi").
Kubernetes consente di eseguire un _piano di controllo_(_control plane_) resiliente, di modo che se un dei _controller_ nativi dovesse fallire, un'altra parte del piano di controllo si occuperà di eseguire quel lavoro.

> Al fine di estendere Kubernetes, si possono avere _controller_ in esecuzione al di fuori del piano di controllo. Oppure, se si desidera, è possibile scriversi un nuovo _controller_. È possibile eseguire il proprio controller come una serie di _Pods_, oppure esternamente rispetto a Kubernetes. Quale sia la soluzione migliore, dipende dalla responsabilità di un dato controller.

Reviewer: s/Pods/Pod/

---
title: Controllers
content_template: templates/concept
weight: 40
---

{{% capture overview %}}

Nella robotica e nell'automazione, un _control loop_ è un un'iterazione senza soluzione di continuità che regola lo stato di un sistema.

Ecco un esempio di un _control loop_ ovvero un _circuito di controllo_: il termostato di una stanza.

Quando viene impostata la temperatura, si definisce attraverso il termostato lo *stato desiderato*. L'attuale temperatura nella stanza è invece lo *stato corrente*. Il termostato agisce per portare lo stato corrente il più vicino possibile allo stato desiderato accendendo e spegnendo le apparecchiature.

{{< glossary_definition term_id="controller" length="short" >}}

{{% /capture %}}

{{% capture body %}}

## Il modello del controller

Un _controller_ monitora almeno una tipo di risorsa registrata in Kubernetes. 

> Questi [oggetti](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects) hanno una proprietà chiamata *spec* che rappresenta lo stato desiderato. Il o i *controller(s)* per quella risorsa sono responsabili di mantenere lo stato corrente il più simile possibile rispetto allo stato desiderato.

Reviewer 2: ```suggestion
Questi [oggetti](/docs/concepts/overview/working-with-objects/kubernetes-objects/#kubernetes-objects) hanno una proprietà chiamata *spec* (specificazione) che rappresenta lo stato desiderato. Il o i *controller(s)* per quella risorsa sono responsabili di mantenere lo stato corrente il più simile possibile rispetto allo stato desiderato.
```

Reviewer 3: (specificazione) --> (specifica)

> Il _controller_ potrebbe eseguire l'azione relativa alla risorsa in questione da sé; più comunemente, in Kubernetes, un _controller_ invia messaggi all'{{< glossary_tooltip text="API server" term_id="kube-apiserver" >}} che produce utili effetti collaterali. Di seguito troverete esempi per questo scenario.

Reviewer 3: I don't like "effetti collaterali" because it has negative acception in Italian, so I'm proposing following translation
```suggestion
Il _controller_ potrebbe eseguire l'azione relativa alla risorsa in questione da sé; più comunemente, in Kubernetes, un _controller_ invia messaggi all'{{< glossary_tooltip text="API server" term_id="kube-apiserver" >}} che a sua volta li rende disponibi ad altri componenti nel cluster. Di seguito troverete esempi per questo scenario.
```

Author: totally agree

{{< comment >}}
Some built-in controllers, such as the namespace controller, act on objects
that do not have a spec. For simplicity, this page omits explaining that

> detail.

Author: Since that's a comment(not showed to the end-user) I thought shouldn't translate. If you think is better do it as well, no problem. I will

> detail.

Reviewer 3: +1 to translate

> detail.

Reviewer 2: (it's not a big deal)

{{< /comment >}}

### Controllo attraverso l'API server

Il {{< glossary_tooltip term_id="job" >}} _controller_ è un esempio di un _controller_ nativo in Kubernetes. I *controllers* nativi gestiscono lo stato interagendo con l'API server presente nel cluster.

> Il Job è una risorsa di Kubernetes che lancia uno o più {{< glossary_tooltip term_id="pod" text="Pods" >}} per eseguire un compito e poi fermarsi.

Reviewer 3: ```suggestion
Il Job è una risorsa di Kubernetes che lancia uno o più {{< glossary_tooltip term_id="pod" text="Pods" >}} per eseguire un task e poi fermarsi.
```
IMO compito is too generic, so I suggest to use "lavoro" or better to stick on task/workload

> (Una volta [programmato](/docs/concepts/scheduling-eviction/), un oggetto _Pod_ diventa parte dello stato desisderato di un dato _kubelet_).

Reviewer 3: ```suggestion
(Una volta che è stato [schedulato](/docs/concepts/scheduling-eviction/), un oggetto _Pod_ diventa parte dello stato desisderato di un dato _kubelet_).
```

Dopo che un nuovo Job è stato creato, lo stato desiderato per quel Job è il suo completamento. Il Job _controller_ fa sì che lo stato corrente per quel Job sia il più vicino possibile allo stato desiderato: creare _Pods_ che eseguano il lavoro che deve essere effetuato attraverso il Job, così che il Job sia prossimo al completamento.

I _controllers_ aggiornano anche gli oggetti che hanno configurato. Ad esempio: una volta che il lavoro relativo ad un dato Job è stato completato, il Job _controller_ aggiorna l'oggetto Job segnandolo come `Finished`.

(Questo è simile allo scenario del termostato che spegne un certo led per indicare che ora la stanza ha raggiungo la temperatura impostata)

### Controllo diretto

A differenza del Job, alcuni _controllers_ devono eseguire delle modifiche a parti esterne al cluster.

> Per esempio, se viene usato un circuito di controllo per assicurare che ci sia un numero sufficiente di {{< glossary_tooltip text="Nodi" term_id="node" >}} nel cluster, allora il _controller_ ha bisogno che qualcosa al di fuori del cluster configuri i nuovi _Nodi_ quando sarà necessario.

Author: not changed due to previous translation agreement

I _controllers_ che interagiscono con un sistema esterno trovano il loro stato desiderato attraverso l'API server, quindi comunicano direttamente con un sistema esterno per portare il loro stato corrente più in linea possibile con lo stato desiderato

(In realtà c'è un _controller_ che scala orizzontalmente i nodi nel cluster. Vedi [Cluster autoscaling](/docs/tasks/administer-cluster/cluster-management/#cluster-autoscaling)).

## Stato desiderato versus corrente {#desired-vs-current}

Kubernetes ha una visione *cloud-native* dei sistemi, ed è in grado di gestire continue modifiche.

> Il cluster viene modificato continuamente durante la sua attività ed il _circuito di controllo_ è in grado di risolvere automaticamente i possibili guasti.

Author: not changed due to previous translation agreement

È utile avere semplici _controllers_ piuttosto che un unico, monolitico, _circuito di controllo_. I _controllers_ possono guastarsi, quindi Kubernetes è stato disegnato per gestire questa eventualità.

{{< note >}}
Ci possono essere diversi _controllers_ che creato o aggiornano lo stesso tipo di oggetti. Dietro le quinte, i _controllers_ di Kubernetes si preoccupano esclusivamente delle risorse (di altro tipo) collegate alla risorsa primaria da essi controllata.

Per esempio, si possono avere _Deployments_ e _Jobs_; entrambe creano _Pods_. Il Job _controller_ non distrugge i _Pods_ creati da un _Deployment_, perché ci sono informazioni (*{{< glossary_tooltip term_id="label" text="labels" >}}*) che vengono usate dal _controller_ per distinguere i _Pods_.
{{< /note >}}

> ## I modi per eseguire i _controllers_ {#running-controllers}

Author: You suggeted to replace "controllers" w/ "controller" all the time in the page/glossary? So, should I replace it here as well?

> ## I modi per eseguire i _controllers_ {#running-controllers}

Reviewer 3: I'm +1 for complying to English grammar when we use English terms (keep the final "s" for plurals)

Kubernetes annovera un insieme di _controllers_ nativi che sono in esecuzione all'interno del {{< glossary_tooltip term_id="kube-controller-manager" >}}. Questi _controllers_ nativi forniscono importanti funzionalità di base.

Il Deployment _controller_ ed il Job _controller_ sono esempi di _controllers_ che vengono forniti direttamente da Kubernetes stesso (ovvero _controllers_ "nativi").

> Kubernetes consente di eseguire un _control plane_ resiliente, di modo che se un dei _controllers_ nativi dovesse fallire, un'altra parte del _control plane_ si occuperà di eseguire quel lavoro.

Author: Fine, no problem. I thought words/expressions that are part of k8s lingo should not being translate, but if you think is better so, no problem, I'll fix it. Should I change "control plane" w/ "piano di controllo" anywhere in page/glossary?

> Kubernetes consente di eseguire un _control plane_ resiliente, di modo che se un dei _controllers_ nativi dovesse fallire, un'altra parte del _control plane_ si occuperà di eseguire quel lavoro.

Reviewer 3: I'm -1 to use "piano di controllo"
"Control plane" (or "control loop") are core concepts of Kubernetes and those terms have very specific meanings. IMO translating those terms might create confusion

> Kubernetes consente di eseguire un _control plane_ resiliente, di modo che se un dei _controllers_ nativi dovesse fallire, un'altra parte del _control plane_ si occuperà di eseguire quel lavoro.

Reviewer 2: To be clear: these are information technology concepts that existed before Kubernetes; in fact, control loops existed before computers were widespread. https://it.wikipedia.org/wiki/Controllore_logico_programmabile#Schede_PID_(Proporzionale_Integrale_Derivativo) uses Italian terms such as “controllori” almost exclusively because that follows the Wikipedia style guide.

_However_ if Italian readers are accustomed to seeing English jargon mixed in, there's not reason to translate everything here. I do like how https://kubernetes.io/pt/docs/concepts/architecture/controller/ provides Portuguese glosses for English terms on their first use. Maybe do that?

> Kubernetes consente di eseguire un _control plane_ resiliente, di modo che se un dei _controllers_ nativi dovesse fallire, un'altra parte del _control plane_ si occuperà di eseguire quel lavoro.

Author: Hi all. I'm a bit confuse about how to proceed. I understood the @sftim point to translate all the world/expression eve if they are part of k8s lingo to be compliant w/ others lan versions (eg. control loop e control plane will be translate including the first time the eg word in parenthesis, while we are not translating "controller"). On the other hand it seems italian k8s community seems more in line w/ @fabriziopandini proposal (https://t.me/kubernetes_it/30623). Please help me on this so I could finalize this issue and use it as guidelines for the next translation. Thanks a lot for you support

> Kubernetes consente di eseguire un _control plane_ resiliente, di modo che se un dei _controllers_ nativi dovesse fallire, un'altra parte del _control plane_ si occuperà di eseguire quel lavoro.

Reviewer 2: I don't speak Italian well, so please dilute my opinions. However, I _recommend_ that you consider that your audience is people who mainly don't already understand Kubernetes.

> Kubernetes consente di eseguire un _control plane_ resiliente, di modo che se un dei _controllers_ nativi dovesse fallire, un'altra parte del _control plane_ si occuperà di eseguire quel lavoro.

Author: Thanks @sftim. I agree on that motivation, I'll try to stick as mach as I could w/ your and Portuguese approach: k8s lingo will be translated adding in parenthesis the en version for the first occurrence.

> Al fine di estendere Kubernetes, si possono avere _controllers_ in esecuzione al di fuori del _control plane_. Oppure, se si desidera, è possibile scriversi un nuovo _controller_. È possibile eseguire il proprio controller come una serie di _Pods_, oppure esternamente rispetto a Kubernetes. Quale sia la soluzione migliore, dipende dalla responsabilità di un dato controller.

Reviewer 3: ```suggestion
Al fine di estendere Kubernetes, si possono avere _controllers_ in esecuzione al di fuori del _control plane_. Oppure, se si desidera, è possibile svilupparre un nuovo _controller_. È possibile eseguire il proprio controller come una serie di _Pods_, oppure esternamente rispetto a Kubernetes. Quale sia la soluzione migliore, dipende dalla responsabilità di un dato controller.
```
